### ISSUE TITLE
[UX] Describe ability to re-order roles

### ISSUE THREAD

**Issue by BWPanda** [backdrop/backdrop-issues#4504]
On `/admin/config/people/roles` you have the ability to re-order the roles, but there's no explanation what this is for. We should add some documentation explaining this.

For the record, Drupal 7 has this help text:

> Roles allow you to fine tune the security and administration of Drupal. A role defines a group of users that have certain privileges as defined on the permissions page. Examples of roles include: anonymous user, authenticated user, moderator, administrator and so on. **In this area you will define the names and order of the roles on your site. It is recommended to order your roles from least permissive (anonymous user) to most permissive (administrator).** To delete a role choose "edit role".
>
> By default, Drupal comes with two user roles:
>
> - Anonymous user: this role is used for users that don't have a user account or that are not authenticated.
> - Authenticated user: this role is automatically granted to all logged in users.

It explains that you can, and suggests how, but doesn't explain *why*... Surely we can do better.

---

PR: https://github.com/backdrop/backdrop/pull/3203

**Comment by BWPanda** [comment]
Also, creating a new role adds it to the end of the list, after 'Administrator'. Assuming Drupal's recommendations about ordering apply to Backdrop too, maybe new roles should be placed between Authenticated and Administrator...?

**Comment by klonos** [comment]
Thanks for raising this @BWPanda 👍 

> To delete a role choose "edit role".

^^ we don't need that 🙂 

> By default, Drupal comes with two user roles

^^ that should be removed, and perhaps those bullet points can be added in the form of description text bellow each respective role.

**Comment by klonos** [comment]
Initial stab at this: https://github.com/backdrop/backdrop/pull/3203

![image](https://user-images.githubusercontent.com/2423362/88847941-74ca6980-d1f0-11ea-915f-658784b50f04.png)

The text is a bit wordy, so please feel free to suggest tweaks. cc @jenlampton @stpaultim @BWPanda and others.

**Comment by jenlampton** [comment]
> maybe new roles should be placed between Authenticated and Administrator...?

Brilliant.

> The text is a bit wordy

Yeah, nobody is going to read this much text. I think separating it into two sections is a good idea @klonos. Let's see if we can get each one down to 1 or 2 sentences a most. How about:

```
Roles provide a way to organize people into groups that have roughly the same duties. Each user account can be given one or more roles, and the Permissions page can be used to grant permissions to those roles (instead of directly to user accounts). 

In this section, you can add or delete roles, define the role names, and change the order. To make the permissions page easier to navigate, we recommended ordering roles from the least permissive (Anonymous) to the most permissive (usually Administrator).
```

Highlights:
* two paragraphs (maybe we move the 2nd one?)
* `users` becomes `user accounts` if we mean accounts, or `people` if we don't.
* removed `you` and make actions passive.
* removed the sentence `This allows updating permissions for entire groups of users, without having to define permissions individually for each user.` since it duplicated what was said in the previous sentence.

Is the bold in the PR just to show what changed? Since changing the order is the least important thing on this page, it shouldn't be emphasized any more than all the other things.

**Comment by klonos** [comment]
> maybe new roles should be placed between Authenticated and Administrator

I think I like this suggestion 👍 

To provide some context for some of the suggestions I'm making bellow, here's a screenshot of what this page looked like in D7:

![image](https://user-images.githubusercontent.com/2423362/88849372-7006b500-d1f2-11ea-81d5-5a015eac3095.png)

Here are some other things that I thought we could improve on this page, or that I could use some feedback for:

1. In D7, there was no way to directly delete a role from this page - the help text stated that `To delete a role choose "edit role"` 👎  ...we've improved the UX in this by providing a "Delete role" option in the operations drop-button 👍 

Now, as a curious user, if I click the drop-button on the first role in the list, I see no delete action available. If I try the next role, it's the same ...so do I keep going, or assume that none of the roles can be deleted? We may know what/why, but we shouldn't be assuming that it is common knowledge that the anonymous and authenticated roles cannot be deleted, whereas the rest can. To that effect, I have changed the vague "required" note next to these special-cased roles with "system role - cannot be deleted", which I think helps this situation (the respective D7 text for this was "locked"). Let me know what you think.

2. I have made the mention to the Permissions page a link, but I think that the UX in D7 was better, with the secondary tabs that allowed to easily switch between the "Roles" and "Permissions" page. I understand that in Backdrop we have moved these pages under "Configuration" instead of directly under "People", but should we consider making those (and the other sections under `Configuration -> User accounts`) tabs? (this will most likely need to be a follow-up) ...or should we alternatively just add some help text at the top of the "Permissions" page, linking back to the roles page?

3. Should we allow descriptions to be added to user roles?

4. There was an entire section of CSS that was trying to put the "Add role" button next to the respective text field. In D7, both the text field and the button were rendered in the same table cell, with `colspan="4"`, but we've moved the button in the "Operations" column in Backdrop. Do we prefer it that way (visually, it makes sense to have it horizontally aligned with the rest of the operations drop-buttons), or should we move it back next to the text field (less distance for the mouse to travel)? I personally don't have a specific preference one way or the other, but as it is, the CSS that was achieving that was not necessary; so I've removed it.

**Comment by jenlampton** [comment]
@klonos these changes are getting way out of scope :) 

> I have made the mention to the Permissions page a link

I did that in my text changes too! (to decrease verbosity)

> or should we alternatively just add some help text at the top of the "Permissions" page, linking back to the roles page?

I like this idea, but yes definitely should be a follow-up issue.

> There was an entire section of CSS

Unrelated, follow-up issue please :)

**Comment by BWPanda** [comment]
I bolded some text in my quote of Drupal's text to show the part I thought was relevent. But happy with whatever text you guys agree to add, as long as it explains the reordering ability.

**Comment by klonos** [comment]
> @klonos these changes are getting way out of scope :)

Of course ...in case you haven't noticed all these years, once I start working on issues, I tend to turn them into "overhaul the UX in this entire admin page". Sorry, and thanks for putting up with my over-excitement 😅 ...will definitely split those to separate follow-ups 👍 

> Is the bold in the PR just to show what changed? Since changing the order is the least important thing on this page, it shouldn't be emphasized any more than all the other things.
> 
>>I bolded some text in my quote of Drupal's text to show the part I thought was relevent.

Oh, I thought it was to emphasize the security best practice implied.

Thanks for the very prompt feedback 🙏 

Working on implementing changes/suggestions...

**Comment by klonos** [comment]
...I still thought that this bit of text had value, because it was explaining the "why" of the previous sentence:

>This allows updating permissions for entire groups of users, without having to define permissions individually for each user.

...but I agree that we should keep things short, so removing.

On the other hand, as I said earlier, I find this bit to be stating the obvious, and of very little value:

>In this section, you can add or delete roles, define the role names, and change their order.

...the reordering ability is already implied by this:

>It is recommended to order your roles from the least permissive (Anonymous) to the most permissive (Administrator)

**Comment by jenlampton** [comment]
> I still think that this bit of text has value, because it is explaining the "why" of the previous sentence:

But there's nothing in here that isn't in the previous sentence, and it was previously said more clearly and concisely: `can be granted to those roles (instead of directly to user accounts).`

> On the other hand, as I said earlier, I find this bit to be stating the obvious, and of very little value:

Okay, let's kill it :)

**Comment by klonos** [comment]
>Maybe we replace `directly` with `individually`?

Yup, that would do it 👍 

PR updated:

- I've changed the help text as per @jenlampton's feedback, and have split it so that there some before and some after the roles table.
- I've restored the removed CSS, and have raised #4506 as a follow-up to either clean that up, or move the "Add role" button back next to the respective text field.

>maybe new roles should be placed between Authenticated and Administrator...?
>
>>Brilliant.

Done. I've implemented this in a simple way, so that new roles are added directly after the authenticated role. We could make that more elaborate, but it'd have to be a separate issue I think.

**Comment by jenlampton** [comment]
> Done. I've implemented this in a simple way, so that new roles are added directly after the authenticated role. We could make that more elaborate, but it'd have to be a separate issue I think.

Is that in a separate PR, or this one for the text change?

**Comment by klonos** [comment]
Same 😅 

Moving it....

**Comment by klonos** [comment]
...unless now that we've added this bit of text it makes it part of this PR:

> ...we recommended ordering roles from the least permissive (Anonymous) to the most permissive (usually Administrator).

**Comment by jenlampton** [comment]
Maybe update the issue title?

**Comment by jenlampton** [comment]
> Done. I've implemented this in a simple way, so that new roles are added directly after the authenticated role. We could make that more elaborate, but it'd have to be a separate issue I think.

This doesn't behave reliably the way it's currently implemented. The first time you add a new role it's fine, because all the weights are the same so it places them alphabetically. But if you then re-order the roles, then try to add another new role, it behaves very strangely.

I think rather than trying to put it one position after the `Authenticated` role, we should try to place it one position before the `Administrator` role. Not only does this more closely match what people will be doing initially (since people would usually add more than one role in increasing order of permissions) but it will also work nicely when people return at a later date and add another new role. 

From an implementation standpoint, we'll need to pull all the current roles by weight, insert the new one at the position where we'd like it, then save all the roles that have new weights. If we are assuming most people leave the `Administrator` role as the most permissive, in most cases this would mean only saving one additional role.  

To place a role after the `Authenticated` role, we'd likely need to save them all. 

Since this is no longer a simple change, I recommend we do this in another issue and keep this one to the text change.

**Comment by klonos** [comment]
> Since this is no longer a simple change, I recommend we do this in another issue and keep this one to the text change. :)

Agreed: #4507

**Comment by klonos** [comment]
Thanks @jenlampton ...I've addressed the low-hanging fruits 👍 ...please have a look at my comments in the PR, and let me know if you insist on the rest of the changes - I'm generally happy to change things according to your suggestions 🙂 

...oh, and I will need help with what seems to be a simple thing: re form element weights when a `theme()` function is called before `backdrop_render_children()` ...setting the weight does not work for me, but you'll most likely know the reason and have a solution for it already.

**Comment by jenlampton** [comment]
I think we should resolve the form element weights issue (I left a new comment on the PR) and push the rewording of `Required` to another follow-up. It's not related to the change requested here and if there's no consensus that it's an improvement, we shouldn't try to shove it in.

**Comment by klonos** [comment]
This is ready for another round please.

**Comment by stpaultim** [comment]
When I went to test, the roles page was set to "show row weights," which was different from expected behavior of "hide row weights." 

I assume, that this is not a change in code but simply the result of someone changing this setting during testing. I changed it back.

**Comment by stpaultim** [comment]
Good work on this change. I'm happy with it. 

One very minor thought that does not need to hold this up, just a simple wording suggestion. The current wording works, but "roughly" feels a tiny bit awkward to me. 

Would "similar duties" be easier for some people to understand than "roughly the same duties"?

WAS: "Roles provide a way to organize people into groups that have roughly the same duties."
SUGGEST: "Roles provide a way to organize people into groups that have similar duties."

**Comment by jenlampton** [comment]
I prefer similar, and it's also less wordy. good suggestion 👍

**Comment by klonos** [comment]
>When I went to test, the roles page was set to "show row weights," which was different from expected behavior of "hide row weights."

This setting is remembered across sessions for the same user, and since everyone is testing in the PR sandboxes using the same admin account, we all get whatever the previous person testing has set this to. So yeah, no relation to the code change. There's only CSS to make sure that the anonymous and authenticated role descriptions are aligned properly independently of whether the table that list the roles has been switched to weight or drag. I imagine that whoever tested last was trying to test this.

> Would "similar duties" be easier for some people to understand than "roughly the same duties"?
>> I prefer similar, and it's also less wordy. good suggestion 👍

Sure 👍 ...PR updated: [`e5567f5` (#3203)](https://github.com/backdrop/backdrop/pull/3203/commits/e5567f5663ddba4183c27a5897da4a2f16036ae4)

**Comment by BWPanda** [comment]
I took a look at the sandbox for this PR, and I have a suggestion: Split off the descriptions for the Anonymous and Authenticated roles into a seperate issue.

This is because I'd like to make the following suggestions which shouldn't hold up this issue and which may need further discussion:
- Put role descriptions in their own column, under the heading 'Description' (as per `/admin/structure/views`) (and make it hidden on smaller screens)
- Add the ability for admins to add descriptions to other roles (i.e. a new 'Description' textfield on `/admin/config/people/roles/edit/[ROLE]`)
- Revert the 'required' text in brackets back to italics

**Comment by BWPanda** [comment]
As for this issue itself, I'd like to please ask that the:

> To make the permissions page easier to navigate, we recommended ordering roles from the least permissive (Anonymous) to the most permissive (usually Administrator).

text be put up the top, as a second paragraph under the first. It looks too hidden/unimportant down the bottom...

**Comment by indigoxela** [comment]
I'm currently not sure about the status of this PR. Is there consensus regarding wording?

**Comment by klonos** [comment]
@indigoxela I believe that there is consensus (mostly 😅 ). Do you have any change to propose?

Can others please provide feedback/opinions re the following suggestions by @BWPanda before I update the PR?

> Split off the descriptions for the Anonymous and Authenticated roles into a seperate issue. ... Put role descriptions in their own column, under the heading 'Description' (as per `/admin/structure/views`) (and make it hidden on smaller screens)

I agree with this ...it makes things consistent with other places, and allows the table to display nicely on mobile 👍 

> Add the ability for admins to add descriptions to other roles (i.e. a new 'Description' textfield on `/admin/config/people/roles/edit/[ROLE]`)

I also agree with this 👍 ...but I personally see no reason not to add these descriptions in the same column as part of this issue now (the above suggestions sound like new features, which means that we'd have to wait till 1.18.0, which is slated for Jan 2021). Once we have all the above, we can move the description text for anonymous and authenticated into their "proper" description properties. The reason why I'm proposing this, is that otherwise we'd need to move those descriptions as bullet points in the main help text at the top of the page (same as it was in D7). The problem I see in that is that this text is already too wordy, and we did a really good job to trim it down; I don't want us to spoil it again.

> Revert the 'required' text in brackets back to italics

Sure. I can do that 👍 

>  I'd like to please ask that the ... text be put up the top, as a second paragraph under the first. It looks too hidden/unimportant down the bottom...

Also can do 👍  ...but I'd like feedback from others on it please.

**Comment by indigoxela** [comment]
> I believe that there is consensus (mostly).

We're sort of being close, but I suspect, we have no real consensus yet.

> Do you have any change to propose?

I don't want to get involved too much in discussions about wording, but to my opinion the text part that explains the meaning of _ordering_ is more important than the one describing the overall meaning of roles.
But I'm not sure what the best place for both of them is.

**Comment by stpaultim** [comment]
Before:

![roles-before](https://user-images.githubusercontent.com/3144571/131777970-9c741e33-f518-4e37-a72b-7823fbc682fc.png)

After:

![roles-after](https://user-images.githubusercontent.com/3144571/131778004-c517e327-d711-4d9a-97e1-909c640ce0de.png)

**Comment by stpaultim** [comment]
Something about the placement of the second paragraph "To make permissions easier to navigate..." that doesn't feel right to me. 

But, would rather see this get merged "as is" than languish for another year. ;-)

**Comment by olafgrabienski** [comment]
I like the side-effect that the page gets a link to the permissions page.

**Comment by docwilmot** [comment]
New PR removes everything but the original issue as in the title: adds help text describing the point of re-ordering.

**Comment by herbdool** [comment]
I agree about @docwilmot approach to keep it simple and focused on the title. I've RTBCed that PR.

I think everything in addition to that should be in a new issue since it needs more discussion.

**Comment by stpaultim** [comment]
I'm in agreement. I was about to mark this "Works for Me." 

Thanks @docwilmot and @herbdool for work today.

**Comment by klonos** [comment]
Thanks everyone 🙏🏼 ...I'll go through this issue and file follow-up issues as required.

**Comment by quicksketch** [comment]
I merged https://github.com/backdrop/backdrop/pull/3888 and closed out https://github.com/backdrop/backdrop/pull/3203.

Thanks everyone for your work and collaboration in working this out!
